File: app/src/main/java/de/kuschku/libquassel/syncables/types/impl/BacklogManager.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.syncables.types.impl;

import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import de.kuschku.libquassel.BusProvider;
import de.kuschku.libquassel.client.Client;
import de.kuschku.libquassel.events.BacklogInitEvent;
import de.kuschku.libquassel.events.BacklogReceivedEvent;
import de.kuschku.libquassel.events.BufferChangeEvent;
import de.kuschku.libquassel.events.ConnectionChangeEvent;
import de.kuschku.libquassel.localtypes.BacklogFilter;
import de.kuschku.libquassel.localtypes.backlogstorage.BacklogStorage;
import de.kuschku.libquassel.message.Message;
import de.kuschku.libquassel.primitives.types.QVariant;
import de.kuschku.libquassel.syncables.types.abstracts.ABacklogManager;
import de.kuschku.util.observables.lists.ObservableComparableSortedList;

import static de.kuschku.util.AndroidAssert.assertNotNull;

public class BacklogManager extends ABacklogManager<BacklogManager> {
    private final Client client;
    private final BacklogStorage storage;
    private final Set<Integer> initialized = new HashSet<>();
    @NonNull
    private final Set<Integer> waiting = new HashSet<>();
    private int waitingMax = 0;
    @IntRange(from = -1)
    private int openBuffer;

    public BacklogManager(Client client, BacklogStorage storage) {
        this.client = client;
        this.storage = storage;
    }

    @Override
    public void requestMoreBacklog(int bufferId, int amount) {
        Message last;
        if (!initialized.contains(bufferId) || null == (last = storage.getUnfiltered(bufferId).last()))
            requestBacklogInitial(bufferId, amount);
        else {
            requestBacklog(bufferId, -1, last.messageId, amount, 0);
        }
    }

    @Override
    public void requestBacklogInitial(int id, int amount) {
        if (waiting.contains(id) || initialized.contains(id))
            return;

        waiting.add(id);
        waitingMax++;
        requestBacklog(id, -1, -1, amount, 0);
    }

    @Override
    public void _requestBacklog(int id, int first, int last, int limit, int additional) {
        // Do nothing, we are on the client
    }

    @Override
    public void _receiveBacklog(int id, int first, int last, int limit, int additional, @NonNull List<Message> messages) {
        assertNotNull(provider);

        storage.insertMessages(id, messages.toArray(new Message[messages.size()]));
        if (messages.size() > 0 && !client.bufferManager().exists(messages.get(0).bufferInfo))
            client.bufferManager().createBuffer(messages.get(0).bufferInfo);
        provider.sendEvent(new BacklogReceivedEvent(id));
        if (id == openBuffer && openBuffer != -1)
            client.bufferSyncer().requestMarkBufferAsRead(openBuffer);
        removeWaiting(id);
    }

    private void removeWaiting(int id) {
        waiting.remove(id);
        initialized.add(id);
        checkWaiting();
    }

    private void checkWaiting() {
        assertNotNull(provider);

        if (client.connectionStatus() == ConnectionChangeEvent.Status.LOADING_BACKLOG) {
            provider.event.postSticky(new BacklogInitEvent(waitingMax - waiting.size(), waitingMax));
            if (waiting.isEmpty()) {
                client.setConnectionStatus(ConnectionChangeEvent.Status.CONNECTED);
            }
        }
    }

    @Override
    public void _requestBacklogAll(int first, int last, int limit, int additional) {
        // Do nothing, we are on the client
    }

    @Override
    public void _receiveBacklogAll(int first, int last, int limit, int additional, @NonNull List<Message> messages) {
        assertNotNull(provider);

        Set<Integer> buffers = new HashSet<>();
        for (Message message : messages) {
            storage.insertMessages(message.bufferInfo.id(), message);
            buffers.add(message.bufferInfo.id());
        }
        for (int id : buffers) {
            provider.sendEvent(new BacklogReceivedEvent(id));
            if (id == openBuffer && openBuffer != -1)
                client.bufferSyncer().requestMarkBufferAsRead(openBuffer);
            waiting.remove(id);
            initialized.add(id);
        }
        checkWaiting();
    }

    @NonNull
    @Override
    public BacklogFilter filter(int id) {
        return storage.getFilter(id);
    }

    @NonNull
    @Override
    public ObservableComparableSortedList<Message> unfiltered(int id) {
        return storage.getUnfiltered(id);
    }

    @NonNull
    @Override
    public ObservableComparableSortedList<Message> filtered(int id) {
        return storage.getFiltered(id);
    }

    @Override
    public void open(int bufferId) {
        assertNotNull(provider);

        openBuffer = bufferId;
        if (bufferId != -1)
            client.bufferSyncer().requestMarkBufferAsRead(bufferId);
        provider.event.postSticky(new BufferChangeEvent());
    }

    @Override
    public int open() {
        return openBuffer;
    }

    @Override
    public void receiveBacklog(@NonNull Message msg) {
        storage.insertMessages(msg);
        if (msg.bufferInfo.id() == openBuffer && openBuffer != -1)
            client.bufferSyncer().requestMarkBufferAsRead(openBuffer);
    }

    @Override
    public int waitingMax() {
        return waitingMax;
    }

    @NonNull
    @Override
    public Set<Integer> waiting() {
        return waiting;
    }

    @Override
    public void _update(Map<String, QVariant> from) {

    }

    @Override
    public void _update(BacklogManager from) {

    }

    @Override
    public void init(@NonNull String objectName, @NonNull BusProvider provider, @NonNull Client client) {
        super.init(objectName, provider, client);
    }
}

File: lib/src/main/java/de/kuschku/libquassel/quassel/syncables/BacklogManager.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2018 Janne Koschinski
 * Copyright (c) 2018 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.quassel.syncables

import de.kuschku.libquassel.protocol.*
import de.kuschku.libquassel.quassel.syncables.interfaces.IBacklogManager
import de.kuschku.libquassel.session.BacklogStorage
import de.kuschku.libquassel.session.Session
import de.kuschku.libquassel.util.compatibility.LoggingHandler.Companion.log
import de.kuschku.libquassel.util.compatibility.LoggingHandler.LogLevel.DEBUG

class BacklogManager(
  private val session: Session,
  private val backlogStorage: BacklogStorage
) : SyncableObject(session, "BacklogManager"), IBacklogManager {
  private val loading = mutableMapOf<BufferId, (List<Message>) -> Boolean>()
  private val loadingFiltered = mutableMapOf<BufferId, (List<Message>) -> Boolean>()

  init {
    initialized = true
  }

  fun updateIgnoreRules() = backlogStorage.updateIgnoreRules(session)

  fun requestBacklog(bufferId: BufferId, first: MsgId = -1, last: MsgId = -1, limit: Int = -1,
                     additional: Int = 0, callback: (List<Message>) -> Boolean) {
    if (loading.contains(bufferId)) return
    loading[bufferId] = callback
    requestBacklog(bufferId, first, last, limit, additional)
  }

  fun requestBacklogFiltered(bufferId: BufferId, first: MsgId = -1, last: MsgId = -1,
                             limit: Int = -1, additional: Int = 0, type: Int = -1, flags: Int = -1,
                             callback: (List<Message>) -> Boolean) {
    if (loadingFiltered.contains(bufferId)) return
    loadingFiltered[bufferId] = callback
    requestBacklogFiltered(bufferId, first, last, limit, additional, type, flags)
  }

  fun requestBacklogAll(first: MsgId = -1, last: MsgId = -1, limit: Int = -1, additional: Int = 0,
                        callback: (List<Message>) -> Boolean) {
    if (loading.contains(-1)) return
    loading[-1] = callback
    requestBacklogAll(first, last, limit, additional)
  }

  fun requestBacklogAllFiltered(first: MsgId = -1, last: MsgId = -1, limit: Int = -1,
                                additional: Int = 0, type: Int = -1, flags: Int = -1,
                                callback: (List<Message>) -> Boolean) {
    if (loading.contains(-1)) return
    loadingFiltered[-1] = callback
    requestBacklogAllFiltered(first, last, limit, additional, type, flags)
  }

  override fun receiveBacklog(bufferId: BufferId, first: MsgId, last: MsgId, limit: Int,
                              additional: Int, messages: QVariantList) {
    val list = messages.mapNotNull<QVariant_, Message>(QVariant_::value)
    if (loading.remove(bufferId)?.invoke(list) != false) {
      log(DEBUG, "BacklogManager", "storeMessages(${list.size})")
      backlogStorage.storeMessages(session, list)
    }
  }

  override fun receiveBacklogAll(first: MsgId, last: MsgId, limit: Int, additional: Int,
                                 messages: QVariantList) {
    val list = messages.mapNotNull<QVariant_, Message>(QVariant_::value)
    if (loading.remove(-1)?.invoke(list) != false) {
      log(DEBUG, "BacklogManager", "storeMessages(${list.size})")
      backlogStorage.storeMessages(session, list)
    }
  }

  override fun receiveBacklogFiltered(bufferId: BufferId, first: MsgId, last: MsgId, limit: Int,
                                      additional: Int, type: Int, flags: Int,
                                      messages: QVariantList) {
    val list = messages.mapNotNull<QVariant_, Message>(QVariant_::value)
    if (loadingFiltered.remove(bufferId)?.invoke(list) != false) {
    log(DEBUG, "BacklogManager", "storeMessages(${list.size})")
    backlogStorage.storeMessages(session, list)
    }
  }

  override fun receiveBacklogAllFiltered(first: MsgId, last: MsgId, limit: Int, additional: Int,
                                         type: Int, flags: Int, messages: QVariantList) {
    val list = messages.mapNotNull<QVariant_, Message>(QVariant_::value)
    if (loadingFiltered.remove(-1)?.invoke(list) != false) {
      log(DEBUG, "BacklogManager", "storeMessages(${list.size})")
      backlogStorage.storeMessages(session, list)
    }
  }

  fun removeBuffer(buffer: BufferId) {
    backlogStorage.clearMessages(buffer)
  }
}
